### PR TITLE
rofiles-fuse: Enable support for setting and getting xattrs

### PR DESCRIPTION
**Pull request by stefanberger**
Enable support for setting and getting xattrs.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

### CODE REVIEW

**Comment by cgwalters** [comment]
Thanks for this patch!  I had actually started a similar patch locally but for some reason I wasn't even seeing the xattr requests go through to the program.  Will give this a try now.

File: src/rofiles-fuse/main.c [hunk 1]
```diff
@@ -529,18 +529,68 @@ callback_access (const char *path, int mode)
...
+get_fullpath (char *fullpath, size_t fullpathsize, const char *path)
+{
+  char buf[32];
+  ssize_t n;
+
+  if (basefd < 0)
+      return -ENOTSUP;
+
+  snprintf (buf, sizeof(buf), "/proc/self/fd/%d", basefd);
+  n = readlink (buf, fullpath, fullpathsize);
```
Comment (cgwalters) [inline src/rofiles-fuse/main.c:542]: Any reason to explicitly `readlink` here versus just constructing the path `/proc/self/fd/%d/user/provided/path` and letting the kernel walk it?

**Review by cgwalters — CHANGES_REQUESTED** [review]

**Comment by cgwalters** [comment]
Thanks for the patch BTW!

File: src/rofiles-fuse/main.c [hunk 2]
```diff
@@ -529,18 +529,68 @@ callback_access (const char *path, int mode)
...
+  if (strlen (fullpath) + strlen (path) + 1 > fullpathsize)
+      return -ENAMETOOLONG;
+  strncat (fullpath, path, fullpathsize);
+
+  return 0;
+}
+
+static int
+is_modifiable_xattr(const char *name) {
+    return (strcmp (name, "user.ima") == 0);
```
[the remark this answers is not in the document]
Comment (stefanberger) [reply src/rofiles-fuse/main.c:556]: I thought this filesystem was a protective mask and there's so much filtering going on that we'd want to be defensive here as well.

**Comment by cgwalters** [comment]
For reference here's the patch I had locally that...wait, hmm it actually seems to be working for me now:

```
From 7bcf67628bf3717cc4c4a284d7a47e3741c9ae87 Mon Sep 17 00:00:00 2001
From: Colin Walters <walters@verbum.org>
Date: Wed, 31 Mar 2021 17:23:52 +0000
Subject: [PATCH] rofiles-fuse: Support xattrs

---
 src/rofiles-fuse/main.c | 34 ++++++++++++++++++++++++++++++----
 1 file changed, 30 insertions(+), 4 deletions(-)

diff --git a/src/rofiles-fuse/main.c b/src/rofiles-fuse/main.c
index 4033caa4..b193dc59 100644
--- a/src/rofiles-fuse/main.c
+++ b/src/rofiles-fuse/main.c
@@ -533,14 +533,28 @@ static int
 callback_setxattr (const char *path, const char *name, const char *value,
                    size_t size, int flags)
 {
-  return -ENOTSUP;
+  PATH_WRITE_ENTRYPOINT (path);
+
+  char buf[PATH_MAX];
+  snprintf (buf, sizeof (buf), "/proc/self/fd/%d/%s", basefd, path);
+
+  if (setxattr (buf, name, value, size, flags) == -1)
+    return -errno;
+  return 0;
 }
 
 static int
 callback_getxattr (const char *path, const char *name, char *value,
                    size_t size)
 {
-  return -ENOTSUP;
+  path = ENSURE_RELPATH (path);
+
+  char buf[PATH_MAX];
+  snprintf (buf, sizeof (buf), "/proc/self/fd/%d/%s", basefd, path);
+
+  if (getxattr (buf, name, value, size) == -1)
+    return -errno;
+  return 0;
 }
 
 /*
@@ -549,8 +563,14 @@ callback_getxattr (const char *path, const char *name, char *value,
 static int
 callback_listxattr (const char *path, char *list, size_t size)
 {
-  return -ENOTSUP;
+  path = ENSURE_RELPATH (path);
+
+  char buf[PATH_MAX];
+  snprintf (buf, sizeof (buf), "/proc/self/fd/%d/%s", basefd, path);
 
+  if (llistxattr (buf, list, size) == -1)
+    return -errno;
+  return 0;
 }
 
 /*
@@ -559,8 +579,14 @@ callback_listxattr (const char *path, char *list, size_t size)
 static int
 callback_removexattr (const char *path, const char *name)
 {
-  return -ENOTSUP;
+  path = ENSURE_RELPATH (path);
+
+  char buf[PATH_MAX];
+  snprintf (buf, sizeof (buf), "/proc/self/fd/%d/%s", basefd, path);
 
+  if (lremovexattr (buf, name) == -1)
+    return -errno;
+  return 0;
 }
 
 struct fuse_operations callback_oper = {
-- 
2.29.2

```

File: src/rofiles-fuse/main.c (hunk 1 above)
Comment (stefanberger) [reply src/rofiles-fuse/main.c:542]: `/proc/self/fd/3` is an actual link and `readlink` seemed good to get the value of it.

File: src/rofiles-fuse/main.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (cgwalters) [reply src/rofiles-fuse/main.c:556]: The main goal of rofiles-fuse is just to make `link()` usable as a mechanism to more safely write to versioned filesystem trees (what ostree does).   It's not explicitly intended as a hard security boundary today for example.  (Hmm, that should be documented)  But on the other hand hopefully we can [just drop it in favor of overlayfs](https://github.com/ostreedev/ostree/issues/2281) at some point soon anyways.

So I don't see a reason to filter the set of xattrs.

File: src/rofiles-fuse/main.c (hunk 1 above)
Comment (cgwalters) [reply src/rofiles-fuse/main.c:542]: Sure, but the kernel will happily traverse it for us just like any other symlink (although it's really a "magic link"); just like you can `ls -al /proc/self/cwd/foo` and `ls` will run directly run e.g.
`statx(AT_FDCWD, "/proc/self/cwd/foo", AT_STATX_SYNC_AS_STAT|AT_SYMLINK_NOFOLLOW, STATX_MODE|STATX_NLINK|STATX_UID|STATX_GID|STATX_MTIME|STATX_SIZE, 0x7ffd789d4280) ...`
without walking the symlink explicitly.

**Comment by cgwalters** [comment]
Is it OK with you if I merge these two, keeping you as author and me as co-author?  Your implementation of `listxattr()` is right for example.

**Comment by cgwalters** [comment]
Proposed merged patch:

```
From 07d143148e0f3356fb427347230a1e6540f89d4c Mon Sep 17 00:00:00 2001
From: Stefan Berger <stefanb@linux.ibm.com>
Date: Wed, 31 Mar 2021 16:41:32 -0400
Subject: [PATCH] rofiles-fuse: Enable support for setting and getting xattrs

Enable support for setting and getting xattrs. Allow modifications
to xattrs only on user.ima xattr.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>
Co-authored-by: Colin Walters <walters@verbum.org>
---
 src/rofiles-fuse/main.c    | 36 ++++++++++++++++++++++++++++++++----
 tests/test-rofiles-fuse.sh | 22 +++++++++++++++++++++-
 2 files changed, 53 insertions(+), 5 deletions(-)

diff --git a/src/rofiles-fuse/main.c b/src/rofiles-fuse/main.c
index 4033caa4..7ace1af0 100644
--- a/src/rofiles-fuse/main.c
+++ b/src/rofiles-fuse/main.c
@@ -533,14 +533,29 @@ static int
 callback_setxattr (const char *path, const char *name, const char *value,
                    size_t size, int flags)
 {
-  return -ENOTSUP;
+  PATH_WRITE_ENTRYPOINT (path);
+
+  char buf[PATH_MAX];
+  snprintf (buf, sizeof (buf), "/proc/self/fd/%d/%s", basefd, path);
+
+  if (setxattr (buf, name, value, size, flags) == -1)
+    return -errno;
+  return 0;
 }
 
 static int
 callback_getxattr (const char *path, const char *name, char *value,
                    size_t size)
 {
-  return -ENOTSUP;
+  path = ENSURE_RELPATH (path);
+
+  char buf[PATH_MAX];
+  snprintf (buf, sizeof (buf), "/proc/self/fd/%d/%s", basefd, path);
+
+  ssize_t n = getxattr (buf, name, value, size);
+  if (n == -1)
+    return -errno;
+  return n;
 }
 
 /*
@@ -549,8 +564,15 @@ callback_getxattr (const char *path, const char *name, char *value,
 static int
 callback_listxattr (const char *path, char *list, size_t size)
 {
-  return -ENOTSUP;
+  path = ENSURE_RELPATH (path);
 
+  char buf[PATH_MAX];
+  snprintf (buf, sizeof (buf), "/proc/self/fd/%d/%s", basefd, path);
+
+  ssize_t n = llistxattr (buf, list, size);
+  if (n == -1)
+    return -errno;
+  return n;
 }
 
 /*
@@ -559,8 +581,14 @@ callback_listxattr (const char *path, char *list, size_t size)
 static int
 callback_removexattr (const char *path, const char *name)
 {
-  return -ENOTSUP;
+  path = ENSURE_RELPATH (path);
+
+  char buf[PATH_MAX];
+  snprintf (buf, sizeof (buf), "/proc/self/fd/%d/%s", basefd, path);
 
+  if (lremovexattr (buf, name) == -1)
+    return -errno;
+  return 0;
 }
 
 struct fuse_operations callback_oper = {
diff --git a/tests/test-rofiles-fuse.sh b/tests/test-rofiles-fuse.sh
index 1e09711c..55816899 100755
--- a/tests/test-rofiles-fuse.sh
+++ b/tests/test-rofiles-fuse.sh
@@ -28,7 +28,7 @@ skip_without_user_xattrs
 
 setup_test_repository "bare"
 
-echo "1..11"
+echo "1..12"
 
 cd ${test_tmpdir}
 mkdir mnt
@@ -75,6 +75,10 @@ fi
 assert_file_has_content err.txt "chown:.*Read-only file system"
 echo "ok failed mutation chmod + chown"
 
+if setfattr -n user.foo -v bar mnt/firstfile-link; then
+    assert_not_reached "set xattr on linked file"
+fi
+
 # Test creating new files, using chown + chmod on them as well
 echo anewfile-for-fuse > mnt/anewfile-for-fuse
 assert_file_has_content mnt/anewfile-for-fuse anewfile-for-fuse
@@ -94,8 +98,14 @@ for i in $(seq 5); do
     chown $(id -u) mnt/newfusedir/test-morenewfuse.${i}
 done
 assert_file_has_content checkout-test2/newfusedir/test-morenewfuse.3 3-morenewfuse-3
+
 echo "ok new content"
 
+setfattr -n user.foo -v bar mnt/anewfile-for-fuse
+getfattr -d -m . mnt/anewfile-for-fuse > out.txt
+assert_file_has_content_literal out.txt 'user.foo="bar"'
+echo "ok new xattrs"
+
 rm mnt/baz/cow
 assert_not_has_file checkout-test2/baz/cow
 rm mnt/baz/another -rf
@@ -152,6 +162,16 @@ firstfile_new_inode=$(stat -c %i checkout-test2/firstfile)
 assert_not_streq "${firstfile_orig_inode}" "${firstfile_new_inode}"
 assert_test_file -f checkout-test2/firstfile
 
+# Test xattr modifications
+copyup_reset
+firstfile_orig_inode=$(stat -c %i checkout-test2/firstfile)
+setfattr -n user.foo -v bar mnt/firstfile
+getfattr -d -m . mnt/firstfile > out.txt
+assert_file_has_content_literal out.txt 'user.foo="bar"'
+firstfile_new_inode=$(stat -c %i checkout-test2/firstfile)
+assert_not_streq "${firstfile_orig_inode}" "${firstfile_new_inode}"
+assert_test_file -f checkout-test2/firstfile
+
 copyup_reset
 firstfile_link_orig_inode=$(stat -c %i checkout-test2/firstfile-link)
 firstfile_orig_inode=$(stat -c %i checkout-test2/firstfile)
-- 
2.29.2

```

**Comment by stefanberger** [comment]
Sounds good.

**Pull request merged**